File: services/nuldoc/nuldoc-src/pages/AboutPage.ts
import GlobalFooter from "../components/GlobalFooter.ts";
import GlobalHeader from "../components/AboutGlobalHeader.ts";
import PageLayout from "../components/PageLayout.ts";
import StaticScript from "../components/StaticScript.ts";
import { Config } from "../config.ts";
import { dateToString } from "../revision.ts";
import { getPostPublishedDate } from "../generators/post.ts";
import { SlidePage } from "../generators/slide.ts";
import {
  a,
  article,
  div,
  elem,
  Element,
  h1,
  h2,
  header,
  img,
  li,
  p,
  section,
  ul,
} from "../dom.ts";

export default async function AboutPage(
  slides: SlidePage[],
  config: Config,
): Promise<Element> {
  return await PageLayout({
    metaCopyrightYear: config.site.copyrightYear,
    metaDescription: "このサイトの著者について",
    metaTitle: `About｜${config.sites.about.siteName}`,
    site: "about",
    config,
    children: elem(
      "body",
      { class: "single" },
      GlobalHeader({ config }),
      elem(
        "main",
        { class: "main" },
        article(
          { class: "post-single" },
          header(
            { class: "post-header" },
            h1({ class: "post-title" }, "nsfisis"),
            div(
              { class: "my-icon" },
              div(
                { id: "myIcon" },
                img({ src: "/favicon.svg" }),
              ),
              await StaticScript({
                site: "about",
                fileName: "/my-icon.js",
                defer: "true",
                config,
              }),
            ),
          ),
          div(
            { class: "post-content" },
            section(
              {},
              h2({}, "読み方"),
              p(
                {},
                "読み方は決めていません。音にする必要があるときは本名である「いまむら」をお使いください。",
              ),
            ),
            section(
              {},
              h2({}, "アカウント"),
              ul(
                {},
                li(
                  {},
                  a(
                    {
                      href: "https://twitter.com/nsfisis",
                      target: "_blank",
                      rel: "noreferrer",
                    },
                    "Twitter (現 𝕏): @nsfisis",
                  ),
                ),
                li(
                  {},
                  a(
                    {
                      href: "https://github.com/nsfisis",
                      target: "_blank",
                      rel: "noreferrer",
                    },
                    "GitHub: @nsfisis",
                  ),
                ),
              ),
            ),
            section(
              {},
              h2({}, "仕事"),
              ul(
                {},
                li(
                  {},
                  "2021-01～現在: ",
                  a(
                    {
                      href: "https://www.dgcircus.com/",
                      target: "_blank",
                      rel: "noreferrer",
                    },
                    "デジタルサーカス株式会社",
                  ),
                ),
              ),
            ),
            section(
              {},
              h2({}, "登壇"),
              ul(
                {},
                ...Array.from(slides)
                  .sort((s1, s2) => {
                    const ta = dateToString(getPostPublishedDate(s1));
                    const tb = dateToString(getPostPublishedDate(s2));
                    if (ta > tb) return -1;
                    if (ta < tb) return 1;
                    return 0;
                  })
                  .map((slide) =>
                    li(
                      {},
                      a(
                        {
                          href:
                            `https://${config.sites.slides.fqdn}${slide.href}`,
                        },
                        `${
                          dateToString(getPostPublishedDate(slide))
                        }: ${slide.event} (${slide.talkType})`,
                      ),
                    )
                  ),
              ),
            ),
          ),
        ),
      ),
      GlobalFooter({ config }),
    ),
  });
}

File: services/nuldoc/lib/nuldoc/pages/about_page.rb
module Nuldoc
  module Pages
    class AboutPage
      def initialize(slides:, staff_records:, config:)
        @slides = slides
        @staff_records = staff_records
        @config = config
      end

      def render
        config = @config
        sorted_slides = @slides.sort_by { |s| GeneratorUtils.published_date(s) }.reverse
        sorted_staff = @staff_records.sort_by(&:sort_date).reverse

        Components::PageLayout.new(
          meta_copyright_year: config.site.copyright_year,
          meta_description: 'このサイトの著者について',
          meta_title: "About｜#{config.sites.about.site_name}",
          site: 'about',
          config: config,
          children: DOM::HTMLBuilder.new.build do
            body class: 'single' do
              render Components::AboutGlobalHeader, config: config
              main class: 'main' do
                article class: 'post-single' do
                  header class: 'post-header' do
                    h1(class: 'post-title') { text 'nsfisis' }
                    div class: 'my-icon' do
                      div(id: 'myIcon') { img src: '/favicon.svg' }
                      render Components::StaticScript,
                             site: 'about',
                             file_name: '/my-icon.js',
                             defer: 'true',
                             config: config
                    end
                  end
                  div class: 'post-content' do
                    section do
                      h2 { text '読み方' }
                      p { text '読み方は決めていません。音にする必要があるときは本名である「いまむら」をお使いください。' }
                    end
                    section do
                      h2 { text 'アカウント' }
                      ul do
                        li do
                          a href: 'https://twitter.com/nsfisis', target: '_blank', rel: 'noreferrer' do
                            text 'Twitter (現 𝕏): @nsfisis'
                          end
                        end
                        li do
                          a href: 'https://github.com/nsfisis', target: '_blank', rel: 'noreferrer' do
                            text 'GitHub: @nsfisis'
                          end
                        end
                      end
                    end
                    section do
                      h2 { text '仕事' }
                      ul do
                        li do
                          text '2021-01～現在: '
                          a href: 'https://www.dgcircus.com/', target: '_blank', rel: 'noreferrer' do
                            text 'デジタルサーカス株式会社'
                          end
                        end
                      end
                    end
                    section do
                      h2 { text '登壇' }
                      ul do
                        sorted_slides.each do |slide|
                          slide_url = "https://#{config.sites.slides.fqdn}#{slide.href}"
                          slide_date = Revision.date_to_string(GeneratorUtils.published_date(slide))
                          li do
                            a href: slide_url do
                              text "#{slide_date}: #{slide.event} (#{slide.talk_type})"
                            end
                          end
                        end
                      end
                    end
                    section do
                      h2 { text 'カンファレンススタッフ' }
                      ul do
                        sorted_staff.each do |record|
                          li do
                            if record.date.is_a?(Range)
                              from_str = Revision.date_to_string(record.date.begin)
                              to_str = Revision.date_to_string(record.date.end)
                              text "#{from_str}〜#{to_str}: #{record.event} (#{record.role})"
                            else
                              date_str = Revision.date_to_string(record.date)
                              text "#{date_str}: #{record.event} (#{record.role})"
                            end
                          end
                        end
                      end
                    end
                  end
                end
              end
              render Components::GlobalFooter, config: config
            end
          end
        ).build
      end
    end
  end
end
